### PR TITLE
Dc move to bo transition

### PR DESCRIPTION
**Pull request by szok**
- [x] Move DC to BO
- [x] Move BO to DC
- [x] Unit tests

### CODE REVIEW

**Comment by mkurek** [comment]
remember about unit tests :)

File: src/ralph/data_center/models/physical.py [hunk 1]
```diff
@@ -435,6 +437,31 @@ def change_rack(cls, instances, request, **kwargs):
...
+    def move_to_bo(cls, instances, request, **kwargs):
+        for instance in instances:
+            back_office_asset = BackOfficeAsset()
+            back_office_asset.__dict__.update(instance.__dict__)
+            back_office_asset.region = Region.objects.get(name='pl')
+            back_office_asset.warehouse = Warehouse.objects.first()
+            back_office_asset.save()
+
+            parents = instance._meta.parents
+            instance._meta.parents = {}
```
Comment (mkurek) [inline src/ralph/data_center/models/physical.py]: write some comment what is going on here and why this workaround is necessary.

File: src/ralph/data_center/models/physical.py [hunk 2]
```diff
@@ -435,6 +437,31 @@ def change_rack(cls, instances, request, **kwargs):
...
+    @transition_action(
+        verbose_name=_('Set as BackOffice Asset'),
+        disable_history=True,
+        disable_save_object=True
+    )
+    def move_to_bo(cls, instances, request, **kwargs):
+        for instance in instances:
+            back_office_asset = BackOfficeAsset()
+            back_office_asset.__dict__.update(instance.__dict__)
+            back_office_asset.region = Region.objects.get(name='pl')
```
Comment (mkurek) [inline src/ralph/data_center/models/physical.py]: region and warehouse should be transition field

File: src/ralph/assets/utils.py [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+    Example:
+        >> instance = DataCenterAsset.objects.get(pk=1)
+        >> back_office_asset = BackOfficeAsset()
+        >> move_parents_models(instance, back_office_asset)
+    """
+    to_obj.__dict__.update(from_obj.__dict__)
+    to_obj.save()
+    parents = from_obj._meta.parents
+    from_obj._meta.parents = {}
+    from_obj.delete()
```
Comment (mkurek) [inline src/ralph/assets/utils.py]: what if there will be some exception during deletion? parent will be empty for all further objects. this should be wrapped in try-finally imo

File: src/ralph/back_office/models.py [hunk 4]
```diff
@@ -506,6 +507,39 @@ def loan_report(cls, instances, request, **kwargs):
             language=kwargs['report_language']
         )
 
+    @classmethod
+    @transition_action(
+        verbose_name=_('Set as DataCenter Asset'),
+        disable_history=True,
+        disable_save_object=True,
+        form_fields={
+            'rack': {
```
Comment (mkurek) [inline src/ralph/back_office/models.py:524]: what about position?

File: src/ralph/back_office/models.py [hunk 5]
```diff
@@ -506,6 +507,39 @@ def loan_report(cls, instances, request, **kwargs):
             language=kwargs['report_language']
         )
 
+    @classmethod
+    @transition_action(
+        verbose_name=_('Set as DataCenter Asset'),
+        disable_history=True,
+        disable_save_object=True,
+        form_fields={
```
Comment (mkurek) [inline src/ralph/back_office/models.py:523]: we should think if model should be changed here as well

File: src/ralph/assets/utils.py [hunk 6]
```diff
@@ -0,0 +1,26 @@
+# -*- coding: utf-8 -*-
+
+
+def move_parents_models(from_obj, to_obj):
```
Comment (mkurek) [inline src/ralph/assets/utils.py:6]: this should be executed in one transaction

**Comment by ghost** [comment]
There should be information about moving in history or something.

File: src/ralph/assets/utils.py [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+    Args:
+        from_obj: From Django model object
+        to_obj: To Django modelo object
+
+    Example:
+        >> instance = DataCenterAsset.objects.get(pk=1)
+        >> back_office_asset = BackOfficeAsset()
+        >> move_parents_models(instance, back_office_asset)
+    """
+    with transaction.atomic():
```
Comment (mkurek) [inline src/ralph/assets/utils.py]: decorator for this function will be enough

File: src/ralph/assets/utils.py [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+        to_obj.__dict__.update(from_obj.__dict__)
```
Comment (mkurek) [inline src/ralph/assets/utils.py]: what about content type of "final" object - is it properly handled here? https://github.com/allegro/ralph/blob/ng/src/ralph/lib/polymorphic/models.py#L176

File: src/ralph/admin/tests/tests_helpers.py [hunk 9]
```diff
@@ -31,3 +37,14 @@ class A():
         a.b = A()
         a.b.name = 'spam'
         self.assertEqual(getattr_dunder(a, 'b__name'), 'spam')
+
```
Comment (ar4s) [inline src/ralph/admin/tests/tests_helpers.py]: use ``@ddt``

File: src/ralph/back_office/tests/test_models.py [hunk 10]
```diff
@@ -271,6 +275,26 @@ def setUp(self):
         messages = FallbackStorage(self.request)
         setattr(self.request, '_messages', messages)
 
+    def test_move_as_data_center_asset(self):
+        bo_asset = BackOfficeAssetFactory()
```
Comment (mkurek) [inline src/ralph/back_office/tests/test_models.py:280]: please add unit test where there is some transition history and attachments attached to asset (we've to make sure that this will be visible from new object too).

File: src/ralph/attachments/views.py [hunk 11]
```diff
@@ -84,3 +75,24 @@ def formset_invalid(self, formset):
...
+
+    def get(self, *args, **kwargs):
+        """
+        All attachments are serving by this view because we need full
+        control (e.g., permissions, rename).
+        """
+        # TODO: respect permissions
+        obj = get_object_or_404(
+            Attachment,
+            id=self.kwargs['id'],
```
Comment (mkurek) [inline src/ralph/attachments/views.py]: why not kwargs['id']? you could use another definition too imo: `def get(self, request, id, filename)`

File: src/ralph/attachments/views.py [hunk 12]
```diff
@@ -84,3 +75,24 @@ def formset_invalid(self, formset):
...
+            original_filename=self.kwargs['filename']
+        )
+        fd = open(obj.file.path, 'rb')
```
Comment (mkurek) [inline src/ralph/attachments/views.py]: you're not closing file descriptor. use context manager here and wrap it in try-except for open-related exceptions (raise 4XX)

File: src/ralph/back_office/models.py [hunk 13]
```diff
@@ -508,6 +510,53 @@ def loan_report(cls, instances, request, **kwargs):
             language=kwargs['report_language']
         )
 
+    @classmethod
+    @transition_action(
+        verbose_name=_('Set as DataCenter Asset'),
```
Comment (mkurek) [inline src/ralph/back_office/models.py]: 'Convert to DataCenter Asset' will fit here better imo

File: src/ralph/back_office/models.py [hunk 14]
```diff
@@ -508,6 +510,53 @@ def loan_report(cls, instances, request, **kwargs):
...
+        }
+    )
+    def move_as_data_center_asset(cls, instances, request, **kwargs):
+        from ralph.data_center.models.physical import DataCenterAsset, Rack  # noqa
+        with transaction.atomic():
+            for i, instance in enumerate(instances):
+                data_center_asset = DataCenterAsset()
+                try:
+                    move_parents_models(instance, data_center_asset)
+                except Exception as e:
```
Comment (mkurek) [inline src/ralph/back_office/models.py]: do we know what exceptions expect here?

File: src/ralph/back_office/models.py [hunk 15]
```diff
@@ -517,7 +566,10 @@ def hostname_assigning(sender, instance, raw, using, **kwargs):
     if getattr(settings, 'BACK_OFFICE_ASSET_AUTO_ASSIGN_HOSTNAME', None):
         if instance.status == BackOfficeAssetStatus.in_progress:
             if instance.pk:
-                bo_asset = BackOfficeAsset.objects.get(pk=instance.pk)
+                try:
+                    bo_asset = BackOfficeAsset.objects.get(pk=instance.pk)
+                except BackOfficeAsset.DoesNotExist:
```
Comment (mkurek) [inline src/ralph/back_office/models.py:574]: add some comment when this exception could occur + add logging here

File: src/ralph/back_office/models.py [hunk 16]
```diff
@@ -517,7 +566,10 @@ def hostname_assigning(sender, instance, raw, using, **kwargs):
...
+                    return
                 if bo_asset.status == BackOfficeAssetStatus.in_progress:
                     return
             logging.info((
```
Comment (mkurek) [inline src/ralph/back_office/models.py]: here should be logger used, not logging

File: src/ralph/lib/transitions/models.py [hunk 17]
```diff
@@ -258,6 +261,8 @@ def run_field_transition(
             action, transition
         ))
         func = getattr(first_instance, action.name)
+        if func.disable_save_object:
```
Comment (ar4s) [inline src/ralph/lib/transitions/models.py]: ??

File: src/ralph/data_center/models/physical.py [hunk 18]
```diff
@@ -439,6 +443,42 @@ def change_rack(cls, instances, request, **kwargs):
         for instance in instances:
             instance.rack = rack
 
+    @classmethod
+    @transition_action(
```
Comment (mkurek) [inline src/ralph/data_center/models/physical.py:480]: this action should be the only action in single transition - add additional param to check for that.

File: src/ralph/lib/transitions/models.py [hunk 19]
```diff
@@ -238,10 +241,9 @@ def run_field_transition(
     """
     Execute all actions assigned to the selected transition.
     """
-    if not isinstance(instances, Iterable):
+    if not isinstance(instances, list):
```
Comment (mkurek) [inline src/ralph/lib/transitions/models.py]: why did you change that?

**Comment by mkurek** [comment]
update functions (form commit comment) should be moved to migration

File: src/ralph/attachments/migrations/0003_auto_20160121_1346.py [hunk 20]
```diff
@@ -0,0 +1,40 @@
...
+        )
+        item.save()
+
+
+def unload_migrate_attachment_item(apps, schema_editor):
+    AttachmentItem = apps.get_model('attachments', 'AttachmentItem')
+    for item in AttachmentItem.objects.all():
+        content_type = item.content_type.model_class().objects.get(
+            pk=item.object_id
+        ).content_type
```
Comment (mkurek) [inline src/ralph/attachments/migrations/0003_auto_20160121_1346.py:26]: this will not work for not-polymorphic objects

File: src/ralph/back_office/models.py (hunk 15 above)
Comment (szok) [inline src/ralph/back_office/models.py:574]: Done

File: src/ralph/data_center/models/physical.py (hunk 18 above)
Comment (szok) [inline src/ralph/data_center/models/physical.py:480]: Done

File: src/ralph/back_office/models.py (hunk 14 above)
Comment (szok) [inline src/ralph/back_office/models.py]: We don't know, unfortunately.

File: src/ralph/lib/transitions/models.py (hunk 19 above)
Comment (szok) [inline src/ralph/lib/transitions/models.py]: Because list it is later modified.

Comment (ghost) [inline src/ralph/lib/transitions/models.py]: This is unclear to me.

Comment (szok) [inline src/ralph/lib/transitions/models.py]: We expect that here is a list of instances not queryset.
Because that list is modified for bulk transitions.
If this is not a list that is that we operate on one object and we can create a new one.

Comment (ghost) [inline src/ralph/lib/transitions/models.py]: Then how it works when `instances` is e.g.: tuple (or any other iterable - is it possible that instances are something else then `list`?)

Comment (szok) [inline src/ralph/lib/transitions/models.py]: Tuple can not be modified.
As I wrote, we need an object that can be modified. The only such a list.

If my solution for the transition is not good, ask for suggestions for a better solution move bo to dc and dc to bo. There must be the possibility of redirects to the new objects from transitions.

Comment (mkurek) [inline src/ralph/lib/transitions/models.py]: So maybe sth like this:

```
if not isinstance(instances, Iterable):  # it only happens where single model instance is passed to this function (old-style transitions)
    instances = [instances]
instances = list(instances)
```

File: src/ralph/lib/transitions/decorators.py [hunk 21]
```diff
@@ -14,6 +14,8 @@ def decorator(func):
         func.run_after = kwargs.get('run_after', [])
         func.help_text = kwargs.get('help_text', '')
         func.precondition = kwargs.get('precondition', lambda instances: {})
+        func.disable_save_object = kwargs.get('disable_save_object', False)
+        func.only_one_action = kwargs.get('only_one_action', False)
```
Comment (mkurek) [inline src/ralph/lib/transitions/decorators.py:18]: add information about these params to transitions documentation

File: src/ralph/lib/transitions/models.py (hunk 19 above)
Comment (szok) [inline src/ralph/lib/transitions/models.py]: This concept is not good.
Because they function inside run_field_transition we can not generate a new list because we lose a reference to it and redirect is will not work.

 I change it to run_field_transition received only a list, then the problem will disappear.

File: src/ralph/lib/transitions/decorators.py (hunk 21 above)
Comment (szok) [inline src/ralph/lib/transitions/decorators.py:18]: Done

File: src/ralph/lib/transitions/models.py [hunk 22]
```diff
@@ -334,10 +338,6 @@ def __new__(cls, name, bases, attrs):
             _transitions_fields[new_class] = fields
             for field in fields:
                 new_class.add_to_class(
-                    'run_{}_transition'.format(field),
```
Comment (szok) [inline src/ralph/lib/transitions/models.py:340]: @ar4s What do you think about delete this? and use only from run_field_transition()

Comment (ar4s) [inline src/ralph/lib/transitions/models.py:340]: :ok:

File: src/ralph/back_office/models.py (hunk 14 above)
Comment (mkurek) [inline src/ralph/back_office/models.py]: so IMO we shouldn't catch exception here - instead of it we should catch every exception in transition action call https://github.com/allegro/ralph/blob/ng/src/ralph/lib/transitions/models.py#L270

File: src/ralph/data_center/tests/test_models.py [hunk 23]
```diff
@@ -159,9 +162,22 @@ def setUp(self):
             parent=self.dc_asset,
         )
 
-    # =========================================================================
-    # slot_no
-    # =========================================================================
```
Comment (mkurek) [inline src/ralph/data_center/tests/test_models.py:164]: why do you remove this?

File: src/ralph/lib/transitions/models.py [hunk 24]
```diff
@@ -68,8 +71,8 @@ def _generate_transition_history(
 
     return TransitionsHistory(
         transition_name=transition.name,
-        content_type=ContentType.objects.get_for_model(instance._meta.model),
-        object_id=instance.pk,
+        content_type=get_content_type_for_model(instance._meta.model),
+        object_id=instance.id,
```
Comment (mkurek) [inline src/ralph/lib/transitions/models.py]: why id, not pk?

File: src/ralph/lib/transitions/tests/test_models.py [hunk 25]
```diff
@@ -14,6 +14,7 @@
     _sort_graph_topologically,
     Action,
     CycleError,
+    run_field_transition,
```
Comment (mkurek) [inline src/ralph/lib/transitions/tests/test_models.py:17]: please test case when exception is raised in some transition action

Comment (szok) [inline src/ralph/lib/transitions/tests/test_models.py:17]: Done

**Comment by mkurek** [comment]
:ok:

File: docs/development/transitions.md [hunk 26]
```diff
@@ -55,6 +55,8 @@ Allowed params for field::
     ``field`` - standard form field, e.g. from ``django.forms``,
     ``condition`` - function wich accept one parameter and return boolean, when condition have be met the field will be shown.
 
+Set ``only_one_action`` to ``True`` if the transition is to be only one action requirements.
```
Comment (xliiv) [inline docs/development/transitions.md]: What does it mean?

File: src/ralph/back_office/models.py [hunk 27]
```diff
@@ -523,10 +569,20 @@ def hostname_assigning(sender, instance, raw, using, **kwargs):
     if getattr(settings, 'BACK_OFFICE_ASSET_AUTO_ASSIGN_HOSTNAME', None):
         if instance.status == BackOfficeAssetStatus.in_progress:
             if instance.pk:
-                bo_asset = BackOfficeAsset.objects.get(pk=instance.pk)
+                try:
+                    bo_asset = BackOfficeAsset.objects.get(pk=instance.pk)
+                except BackOfficeAsset.DoesNotExist:
+                    # Can not assign a new hostname, because there are
+                    # not yet saved the object.
+                    logger.debug(
```
Comment (szok) [inline src/ralph/back_office/models.py]: use info

File: src/ralph/back_office/models.py [hunk 28]
```diff
@@ -514,6 +515,51 @@ def loan_report(cls, instances, request, **kwargs):
             language=kwargs['report_language']
         )
 
+    @classmethod
+    @transition_action(
+        verbose_name=_('Convert to DataCenter Assett'),
```
Comment (szok) [inline src/ralph/back_office/models.py]: assett

**Comment by ghost** [comment]
:ok:
